stdlib/in_channel.ml

```diff
@@ -30,6 +30,22 @@ let stdin = Stdlib.stdin
 let open_bin = Stdlib.open_in_bin
 let open_text = Stdlib.open_in
 let open_gen = Stdlib.open_in_gen
+
+let with_open openfun s f =
+  let ic = openfun s in
+  match f ic with
+  | r -> Stdlib.close_in ic; r
+  | exception e -> Stdlib.close_in_noerr ic; raise e
```

Comment:
Any reason why not to use the shiny `Fun.protect` ? It will do nice things for you like preserve the backtrace etc.

Reply:
I wanted to keep the dependencies to a minimum. By the way, I believe the above definition preserves the backtrace of `e`.

Reply:
I honestly don't know, but `Fun.protect` seems to be more serious about it.

Also I'm a bit unsure of what to think about the asymetry in closing behaviour between a result or raising exception. 

I would tend to do `close_in_noerr` in both cases (like a use of `Fun.protect` would do).

Reply:
I don't have a strong opinion against using `Fun.protect`. To be honest, this is how I always wrote the function before the advent of `Fun` and didn't think much of it. As for the difference between `close_in` vs `close_in_noerr`: in my mind, one uses `close_in_noerr` to avoid masking `e` in the exception case, but I don't see why errors while closing should be ignored in general.

Reply:
> As for the difference between `close_in` vs `close_in_noerr`: in my mind, one uses `close_in_noerr` to avoid masking `e` in the exception case, but I don't see why errors while closing should be ignored in general.

For me the contract of `with_*` brackets/`Fun.protect`ed function is that they fully handle a ressource and its cleanup for you. 

Now with the current implementation you handle it in some of the cases but not in other ones. What am I supposed to do about this `close_in` error ? It's not really my business, you were supposed to handle that. What happened to the underlying `fd`, did it leak ? 

When that `close_in` raises it poses more questions than it solves in my opinion.

(Aside, this shows yet another advantage of using explicit errors via `result`, they allow to clearly distinguish error scopes in these cases. E.g. with a sig like `with_open : string -> (t -> 'a) -> ('a, syserr) result`, you clearly delineate the bracket scope error from the client ones, which can use either a result for `'a` or raise exceptions).

Reply:
I'm convinced :) I switched to `Fun.protect`.